### ISSUE TITLE
UX smoother: can OSARA's hard coded use of the Applications key also be extended to Shift+F10?

### ISSUE THREAD

**Issue by ScottChesworth** [jcsteh/osara#432]
Getting newbies on laptops remapped with an Applications key is a frequent topic across our support groups. Given how soon into the journey this trip hazard tends to present itself (activating MIDI inputs for example), I thought it'd be worth asking whether OSARA can add Shift+F10 to the code that currently hacks context menus where necessary?

Thanks in advance if there's anything you can do.

**Comment by jcsteh** [comment]
Yeah, been thinking about this for a while. There's a catch, though. time second context menu is currently shift+apps. We can't do shift+shift+f10. :) Thankfully, there's no control+apps yet, so we could do control+shift+f10, but this would make the shift+f10 variants inconsistent with the apps variants.

**Comment by ScottChesworth** [comment]
Isn't the time second context menu accessible via the "OSARA: Show second context menu (depending on focus)" action nowadays, though? I'm only proposing Shift+F10 be made available in places where those three context menu actions don't yield any joy for newbies. Setting inputs and outputs in MIDI Devices category of Prefs and the Delete buttons for sends and receives in the I/O window are the first two areas springing to mind. If there are other places, you'd likely know better than me.

**Comment by jcsteh** [comment]
It is, yes, but I'm guessing having to bind actions to those keys isn't trivial for newbies either. Though I guess we could do that in the default key map. If I recall correctly, you can't bind stuff to apps, but you can hopefully bind to shift+f10. Though that might break context menus in some other areas outside the track view.

**Comment by ScottChesworth** [comment]
I just tried. The UX was a bit lumpy (had to dismiss an unexpected context menu when binding Apps and Control+Apps), but the assignments did seem to stick, so we could do those as default mappings.

Annoyingly though, that can't extend to all three modifiers on Shift+F10, because Alt+Shift+F10 is currently "OSARA: Reset Peak Watcher for second track". Doh!

**Comment by jcsteh** [comment]
My guess is that if you bind those, context menus elsewhere might break. For example, try renaming a track and accessing the text box context menu.

**Comment by ScottChesworth** [comment]
Good news, it's weirder than that. :)
Context menus still work everywhere as expected. However, the actions assigned to Apps, Control+Apps and Alt+Apps aren't reported in Shortcut Help. That made me think that perhaps the assignments hadn't stuck in some way. They are detected when I use the Find shortcut button in the actions list though.
Colour me confused.

**Comment by jcsteh** [comment]
OSARA is almost certainly overriding those action bindings with its own keyboard hook, so that makes sense. I guess the question is whether they work properly with OSARA's special handling code removed.

**Comment by ScottChesworth** [comment]
Is there something I can comment out to temporarily override the keyboard hook? Happy to do that, rebuild and test locally if that'd be helpful.

**Comment by jcsteh** [comment]
> The second context menu is currently shift+apps.

I was wrong about this anyway. It's apps, control+apps and alt+apps; there's no shift+apps.

> The UX was a bit lumpy (had to dismiss an unexpected context menu when binding Apps

Interesting. I couldn't bind apps at all. Control+apps and alt+apps worked, but not just apps. Still, if at least someone can bind them, we can include them in a key map.

> Annoyingly though, that can't extend to all three modifiers on Shift+F10, because Alt+Shift+F10 is currently "OSARA: Reset Peak Watcher for second track". Doh!

I guess we can do alt+control+shift+f10 or Windows+shift+f10.

I do wonder whether binding apps will break opening the context menu in the MIDI editor. Currently, you can do that with apps, but I suspect binding apps in the main section will break this.

I guess the question from here is whether to handle the OSARA: Show blah context menu bindings as we currently do or whether to get them into the key map instead. Given that we're probably going to end up breaking consistency based on a key map conflict (alt+shift+f10), it does feel like it'd be better to bind these in the key map if we can. On the other hand, if that's going to break the MIDI editor, that's a problem.

**Comment by jcsteh** [comment]
Shift+f10 is supported for ListView items in 3f80f0e. I haven't done this for the send menu yet because the track view context menus are currently handled by the same block of key hook code as the send menu, so how I refactor that depends on the decisions concerning the track view context menus.

**Comment by ScottChesworth** [comment]
> I couldn't bind apps at all. Control+apps and alt+apps worked, but not just apps.
> 
I had to use the Special key button to bind Apps without modifiers. Sorry, forgot to mention that.
> 
> I guess we can do alt+control+shift+f10 or Windows+shift+f10.
> 
Should we make the mapping for third context menu Control+Alt+Apps for consistency then?
> I do wonder whether binding apps will break opening the context menu in the MIDI editor. Currently, you can do that with apps, but I suspect binding apps in the main section will break this.
> 
Irrespective of whether it breaks context menu access in the MIDI editor or not, I should probably hassle Cockos about getting Alt to work in there. It looks like a standard menu bar and already behaves as one over on Mac, so if that can be reflected in Windows then we've ironed out an instance of disparity and the change of Apps behaviour won't matter.

**Comment by ScottChesworth** [comment]
> I haven't done this for the send menu yet because the track view context menus are currently handled by the same block of key hook code as the send menu, so how I refactor that depends on the decisions concerning the track view context menus.
> 
Gotcha, I'll float that in the keymap group now then.
All three context menus are working nicely mapped to Shift+F10, Control+Shift+F10 and Control+Alt+Shift+F10 in the TCP here. Shift+F10 also works to select MIDI inputs and outputs in Prefs. Apps key still works as expected inside the MIDI editor, but I've also emailed Cockos about fixing up the Alt key in there just for consistency.

Question: the new mappings around Shift+F10 report in Shortcut Help. At the moment, the context menu actions don't report when mapped to Apps key with or without modifiers though. Is there anything you can do to get the Apps key reporting?

**Comment by GarthHum** [comment]
It’s probably worth mirroring these on Mac as well. 
Shift-F10, Shift-Command-F10, and Shift-Option-Command-F10 for the 3 context menus

**Comment by jcsteh** [comment]
Concerning the apps key, we have a bit of a chicken and egg situation. It's not working in shortcut help because OSARA's key hook is intercepting that key before REAPER can process it. However, if I remove that, it'll break things until we get the new key map. If you can email me the key map based on latest OSARA with just those keys added, I'll push a PR with that added and the key hook modified accordingly for broader testing.

Regarding control+alt+apps, while I like consistency, I worry about yet another backwards compat break here. We're also making that key harder to hit for the sake of keyboards without an apps key. I wonder whether this is a case where we just have to let inconsistency stand in service of pragmatism. That said, you're probably the one that's going to end up dealing with the fallout of such a change, so I won't fight if you want it changed.

**Comment by tbdalgaard** [comment]
No, please let the Mac keys stay since this might break due to people having a touch bar. Haven't tested this myself yet, but please hold the horses until someone has tested this

**Comment by jcsteh** [comment]
> I had to use the Special key button to bind Apps without modifiers. Sorry, forgot to mention that.

I did try that and I couldn't even get that to work. Weird. However, my laptop doesn't have an apps key, so I have a weird NVDA gesture hack to get around that rather than using SharpKeys for... reasons. That might be causing weirdness.

**Comment by jcsteh** [comment]
> my laptop doesn't have an apps key, so I have a weird NVDA gesture hack to get around that rather than using SharpKeys for... reasons. That might be causing weirdness.

That was indeed the problem. I tried with an external keyboard and was able to bind. However...

> At the moment, the context menu actions don't report when mapped to Apps key with or without modifiers though. Is there anything you can do to get the Apps key reporting?

I removed the OSARA key hook command related to this. Unfortunately, even with that removed, the apps key doesn't behave properly even once it is bound. REAPER just acts as if it doesn't get the key at all. I knew there was a reason I ended up using a keyboard hook instead of binding this to an action all those years ago. :)

So, we still need special code for the apps key in the track view unfortunately, but we can bind the shift+f10 variants to actions normally.
